### ISSUE TITLE
If model has incrementing id "save()" won't invalidate

### ISSUE THREAD

**Issue by saernz** [GeneaLabs/laravel-model-caching#242]
**Describe the bug**
If you have table migrations for your models that use `$table->increments('id');` rather than just a `$table->integer('id')->unsigned();` the library fails to invalidate the cache on save().

This is because a code path is followed which falls back to the base query. The code path is: Illuminate\Database\Eloquent\Model > save() > performInsert() > **insertAndSetId()** > Illuminate\Database\Eloquent\Builder::insertGetId() > __call > toBase().

The problem happens when the Eloquent model returns true for `getIncrementing()` in the `performInsert()` function of the Eloquent Model. Because `CachedBuilder` from the library doesn't override `insertAndSetId()` to perform a `checkCooldownAndFlushAfterPersisting()` the cache is never flushed and this results in dirty reads from the cache.

**Eloquent Query**
For some reason I could only replicate this in testing. My test is pretty much the below structure. The library was actually behaving when I'm using my development environment but when I run my tests in the testing environment any model which used incremental ids would fail if it did multiple writes and then tested the returned results.

```php
        $this->assertEquals(0, SomeModel::get()->count());
        $this->somePostWhichCreates($data);
        $this->assertEquals(1, SomeModel::get()->count());
```

**Environment**
 - PHP: 7.2
 - Laravel: v5.8.15
 - Model Caching: 0.4.14

**Work around**
At the moment the work around I have found to fix this issue is to use the "BaseModel" setup method and then add the following method to the base model:

```php
    /**
     * Insert the given attributes and set the ID on the model.
     *
     * @param  \Illuminate\Database\Eloquent\Builder  $query
     * @param  array  $attributes
     * @return void
     */
    protected function insertAndSetId(Builder $query, $attributes)
    {
        $this->checkCooldownAndFlushAfterPersisting($this);
        parent::insertAndSetId($query, $attributes);
    }
```

Unfortunately I cannot leave a link to the code base as it's a private repo and my company wouldn't be happy.

For some reason I've only noticed this in testing. Though when I look at the code it seems to look like this may be a bug. I might need someone with better knowledge of the code to deny or confirm this.

**Comment by mikebronner** [comment]
@saernz Thanks for reporting this. I will look into this soon.

**Comment by mikebronner** [comment]
Likely related to #225

**Comment by mikebronner** [comment]
@saernz Are you overriding your model's `created`/`deleted`/`saved` static methods? That's what triggers cache invalidation. I have not been able to replicate your issue as described.

**Comment by saernz** [comment]
Ahh I have found out what I have done, and I must apologise if I've caused any wasted time.

I did not realise that by using Event::fake() in my tests that this also stops model created/deleted/saved events as well! Some of my tests contain:
```php
    /**
     * Setup the test environment.
     *
     * @return void
     */
    protected function setUp(): void
    {
        parent::setUp();
        Event::fake();
    }
```

which is causing the problem. Very sorry for wasting your time. When I remove Event::fake() it seems to fix the problem. I was also confused when looking at the code as I didn't realise Eloquent also boots it's traits and I thought the bootCachable was dead code, which it obviously isn't.
